pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -0,0 +1,308 @@
+package tenantfederation
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	"github.com/weaveworks/common/user"
+
+	"github.com/cortexproject/cortex/pkg/tenant"
+)
+
+const defaultTenantLabel = "__tenant_id__"
+
+// NewQueryable returns a queryable that iterates through all the tenant IDs
+// that are part of the request and aggregates the results from each tenant's
+// Querier by sending of subsequent requests.
+// The result contains a label tenantLabelName to identify the tenant ID that
+// it originally resulted from.
+// If the label tenantLabelName is already existing, its value is overwritten
+// by the tenant ID and the previous value is exposed through a new label
+// prefixed with "original_". This behaviour is not implemented recursively
+func NewQueryable(upstream storage.Queryable) storage.Queryable {
+	return &mergeQueryable{
+		upstream: upstream,
+	}
+}
+
+type mergeQueryable struct {
+	upstream storage.Queryable
+}
+
+// Querier returns a new mergeQuerier aggregating all tenants into the result
```

Comment:
Godoc style comments should end with a full stop. Looks like the full stop needs to be added to a few godoc comments in here.

Reply:
:+1: I also rephrase it a bit